### PR TITLE
XmlDataContractSerializerInputFormatter Cleanup

### PR DESCRIPTION
**Pull request by tonysneed**
Commit dcfa16d:
Refactored CreateDataContractSerializer method to return DataContractSerializer instead of XmlObjectSerializer, as discussed in #1871. Also renamed method to CreateSerializer and fixed up the xml comments.

Commit 47b80ca:
Added virtual GetSerializableType method to provide extensibility point for creating xml that can be deserialized into a wrapped object, similar to the same method in XmlDataContractSerializerOutputFormatter. Relates to #1872.  Placed call to SerializableErrorWrapper.CreateSerializableType in the GetSerializableType method and removed it from the CreateSerializer method.

### CODE REVIEW

**Comment by tonysneed** [comment]
PR #1871 should be merged before this one.

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs [hunk 1]
```diff
@@ -101,12 +121,23 @@ protected virtual XmlReader CreateXmlReader([NotNull] Stream readStream)
         }
 
         /// <summary>
-        /// Called during deserialization to get the <see cref="XmlObjectSerializer"/>.
+        /// Gets the type of the object to be serialized.
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: nit: `Gets the type of the object to be deserialized to` ?

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs [hunk 2]
```diff
@@ -101,12 +121,23 @@ protected virtual XmlReader CreateXmlReader([NotNull] Stream readStream)
...
+        /// </summary>
+        /// <param name="declaredType">The declared type.</param>
+        /// <returns>The type of the object to be serialized.</returns>
+        protected virtual Type GetSerializableType([NotNull]Type declaredType)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: nit: space between NotNull and Type

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs [hunk 3]
```diff
@@ -119,16 +150,16 @@ private object GetDefaultValueForType(Type modelType)
             return null;
         }
 
-        private Task<object> ReadInternal(InputFormatterContext context)
+        private Task<object> ReadInternalAsync(InputFormatterContext context)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: nice

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/XmlDataContractSerializerInputFormatterTest.cs [hunk 4]
```diff
@@ -359,6 +366,118 @@ public void XmlDataContractSerializerFormatterThrowsWhenMaxDepthIsBelowOne()
             Assert.Equal("Test Error 3", serializableError["key2"]);
         }
 
+        [Fact]
+        public async Task XmlDataContractSerializerFormatterThrowsWhenNotConfiguredWithRootName()
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/XmlDataContractSerializerInputFormatterTest.cs]: You can remove the prefix `XmlDataContractSerializerFormatter` as its redundant(as the file name is grouping the tests) and makes it difficult to read

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs [hunk 5]
```diff
@@ -119,16 +150,16 @@ private object GetDefaultValueForType(Type modelType)
...
 
             using (var xmlReader = CreateXmlReader(new DelegatingStream(request.Body)))
             {
-                var xmlSerializer = CreateDataContractSerializer(type);
-                var deserializedObject = xmlSerializer.ReadObject(xmlReader);
-                deserializedObject = SerializableErrorWrapper.UnwrapSerializableErrorObject(type, deserializedObject);
+                var type = GetSerializableType(context.ModelType);
+                var dataContractSerializer = CreateSerializer(type);
+                var deserializedObject = dataContractSerializer.ReadObject(xmlReader);
+                deserializedObject = SerializableErrorWrapper.UnwrapSerializableErrorObject(context.ModelType, deserializedObject);
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: should be `type` instead of `context.ModelType`

**Comment by kichalla** [comment]
Looks good to me! :shipit:

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs (hunk 2 above)
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: Could you please do this kind of cleanup for `XmlSerializerInputFormatter` too?

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs (hunk 1 above)
Comment (tonysneed) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: Oh yes, copied this verbatim from XmlDataContractSerializerOutputFormatter.  It should instead say: "Gets the type to which the XML will be deserialized."

File: src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs (hunk 2 above)
Comment (tonysneed) [inline src/Microsoft.AspNet.Mvc.Xml/Formatters/XmlDataContractSerializerInputFormatter.cs]: Great idea! Glad to do that.

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/XmlDataContractSerializerInputFormatterTest.cs (hunk 4 above)
Comment (tonysneed) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/XmlDataContractSerializerInputFormatterTest.cs]: I did that in a separate PR: #1887 commit 97c258e.

**Comment by tonysneed** [comment]
I implemented corrections to XmlDataContractSerializerInputFormatter and applied cleanup to XmlSerializerInputFormatter.

**Comment by yishaigalatzer** [comment]
@kichalla please merge

**Comment by kichalla** [comment]
@tonysneed : Could you please squash your commits so that its easier for us to merge?

**Comment by tonysneed** [comment]
No problem, happy to squash the commits down to one.

**Comment by tonysneed** [comment]
Created a new PR #1917 with the original commits condensed to a single commit.

**Comment by kichalla** [comment]
Merged with aspnet/dev : c6de763a6cc262192dbe0c3904047da52df86f63

**Pull request closed without merging**